crates/chain/Cargo.toml

```diff
@@ -13,18 +13,19 @@ readme = "../README.md"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-bitcoin = { version = "0.29" }
+bitcoin = { version = "0.29", default-features = false }
 serde_crate = { package = "serde", version = "1", optional = true, features = ["derive"] }
 
 # Use hashbrown as a feature flag to have HashSet and HashMap from it.
 # note version 0.13 breaks outs MSRV.
-hashbrown = { version = "0.12", optional = true,  features = ["serde"] }
-miniscript = { version = "9.0.0", optional = true  }
+hashbrown = { version = "0.12", optional = true, features = ["serde"] }
+miniscript = { version = "9.0.0", optional = true, default-features = false }
 
 [dev-dependencies]
 rand = "0.8"
 
 [features]
 default = ["std", "miniscript"]
-std = []
+std = ["bitcoin/std", "miniscript/std"]
+no-std = ["hashbrown", "bitcoin/no-std", "miniscript/no-std"]
```

Comment:
Why does rust bitcoin and rust miniscript make no-std a feature :S

I am looking at the source of rust miniscript and rust bitcoin and it looks like the `no-std` feature doesn't do anyhing (absence of `std` makes it no_std)`

Let's not add this feature if this is the case.

Reply:
As I understand it the `no-std` feature turns on some dependencies that are needed to use `rust-bitcoin` and `rust-miniscript`, in a no std environment. In particular it looks like it's enabling some `alloc` features in other dependencies. For this PR I don't see another way besides the downstream users manually enabling what's needed, the `no-std` feature is more of a convenience to add what's needed, the "hashbrown" and the bitcoin and miniscript `no-std` features. Since there's no way to enable the `no-std` dependencies due to the absence of `std` I don't see another way.

Reply:
> Since there's no way to enable the `no-std` dependencies due to the absence of `std` I don't see another way. 

Ok so I see that `hashbrown` needs to be enabled to have `HashMap`. I think what we do in `bdk_chain` is the right way to solve this specific problem: Use `BTreeMap` instead unless `hashbrown` feature is enabled which makes it faster by using an actual hashbrown hashmap. If that's too offensive then just make `hashbrown` a mandatory dependency. That seems way less harmful than this feature flag.

I think adding this feature just to enable features in another crate is not a good solution:
- Crates depending on `bdk` will have to do the same thing and so the `no-std` thing spreads everywhere.
- It means `--all-features` enabling `std` and `no-std`  which just seems awful.

I think just telling people that they have to manually depend on `bitcoin` and or `miniscript` and enable their `no-std` feature manually is probably the right way to go for now. I hope we can remove `no-std` upstream.

Reply:
@tcharding gave us some good advice:

![image](https://user-images.githubusercontent.com/3734083/226513283-100a72b4-b6c5-4624-b869-05b970cdf5c8.png)

we can just always enable the `no-std` feature and optionally enable std.

Reply:
Unfortunately you'll get a dependency on `hashbrown` that is unneeded in "std" if you do the same thing for `miniscript`.

Reply:
I had a quick look and I rekon it should work for miniscript also.